game-core/src/main/java/games/strategy/engine/chat/ChatMessagePanel.java

```diff
@@ -222,9 +230,13 @@ public void addMessage(final String message, final PlayerName from) {
     addMessageWithSound(message, from, SoundPath.CLIP_CHAT_MESSAGE);
   }
 
-  /** thread safe. */
   @Override
-  public void addMessageWithSound(final String message, final PlayerName from, final String sound) {
+  public void addSlapMessage(final String message, final PlayerName from) {
+    addMessageWithSound(message, from, SoundPath.CLIP_CHAT_SLAP);
+  }
+
+  private void addMessageWithSound(
```

Comment:
I think we're still left with a SRP violation.

What do you think of making the private method to be focused on playing a sound?

eg:

```
  public void addSlapMessage(final String message, final PlayerName from) {
    addChatMessage(message, from);
    playSound(from, SoundPath.CLIP_CHAT_SLAP);
  }
```

Reply:
Good point, I'm not sure if this will have unintended consequences though:
If this listener is called on the EDT, this operation is blocking, otherwise it's executed asynchronously. So there are a handful of options.
I'd assume that the nio listeners are not running on the EDT for simplicity, otherwise this problem would be trivial to solve. Let me know if you have any insights about which Thread invokes this listener.

The options:
1. Run `SwingAction.invokeNowOrLater` directly in `addMessage` and `addSlapMessage`, and turn `addMessageWithSound` into 2 blocking methods.
2. Turn `addMessageWithSound` into 2 methods that each invoke `SwingAction.invokeNowOrLater` separately and execute them one by one. This way the sound might not be playing synchronized with the actual chat event.
3. Like 2., but it's guaranteed to be asynchronous even when executed on the EDT.

**However there's a problem with all of those approaches:**
The flood control check is done asynchronously as well, and obviously the sound should only be played if floodControl allows it. I believe the advertised "thread-safety" is mainly coming from the fact that everything is executed on the EDT, so I wouldn't feel confident moving those checks to the current thread :(
Thoughts?

Reply:
It might not be thread safe, if flood control is checked on 'n' parallel threads, are all the needed values marked as volatile to guarantee each thread will see the updated values from other threads?

I think the split could be done by moving the SwingAction async logic to the top level, so we then just invoke two methods from within that thread rather than having the lower-level methods do so.

The flood control check IMO would be better also moved to a top layer, fits better with "imperative shell, functional core" design pattern (in other words, control flow logic at a higher level calling functional components). From another perspective, the 'addChatMessage' method is incorrectly named as it does not always do that, it checks flood control and then has conditions based on that.

Another benefit to moving control-flow logic to an imperative shell is it's easier to understand larger sections of the code. Currently if we inlined all the code, it would be all the same for understanding. Today, you have to go through all the lines of every method invoked to know what will happen, the method migth be a no-op, it might not be, to know you have to look at them. One trick to maintainable code is to divide and conquer what someone needs to know about. For example "playSoundMethod" that maybe plays a sound, you have to read it all. But if you see there is a condition on the method call, and you know the condition is false, then that's an entire private method you do not have to read as you know it's not executed (rather than executed but resulting in a no-op).